federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,956 @@
+	// write back to federation apiserver
+
+	// TODO to check LB info equals
+	//if !api.LoadBalancerStatusEqual(previousState, &service.Status.LoadBalancer) { // TODO: quinton: only interested in the DNS record here.
+	//	// Not even sure it needs updating - check.
+	//} else {
+	_, err := s.federationClient.Core().Services().Update(serviceInMap)
+	if err != nil {
+		return fmt.Errorf("fail to update federation service")
+	}
+	//	glog.V(2).Infof("Not persisting unchanged LoadBalancerStatus to registry.") // TODO: quinton: check/fix this.
+	//}
+
+	return nil, notRetryable
+}
+
+// Returns whatever error occurred along with a boolean indicator of whether it
+// should be retried.
+func (s *ServiceController) ensureDNSRecordsExist(key string, fedService, appliedState *v1.Service) (error, bool) {
+	// Clone federation service, and create them in underlying k8s cluster
+	clone, err := conversion.NewCloner().DeepCopy(fedService)
+	if err != nil {
+		return nil, err
+	}
+	service, ok := clone.(*v1.Service)
+	if !ok {
+		return nil, fmt.Errorf("Unexpected service cast error : %v\n", service)
+	}
+	// to generate service name, we need a api.ObjectMeta instead of v1
+	meta := &api.ObjectMeta{}
+	meta.GenerateName = service.ObjectMeta.Name + "-"
+	service.Name = meta.Name
+	service.GenerateName = fedService.Name
+	service.ResourceVersion = ""
+
+	api.GenerateName(api.SimpleNameGenerator, meta)
+	// TODO: to generate labels based on hash
+
+	// handle available clusters one by one
+	for clusterName, client := range s.clusterClientMap {
+		go s.processServiceForCluster(key, clusterName, service, appliedState, client)
+	}
+}
+
+func retrieveExternalIPs(client clientset.Clientset, service v1.Service) {
+	// TODO: sleep and pull service info to check LB changes? or create k8s cluster service watch when create clusterClientMap
+	for i := 0; i < clientRetryCount; i++ {
+		time.Sleep(clientRetryInterval)
+		svc := getServiceByLabel(client, service)
+		if svc != nil {
+			return svc.Spec.ExternalIPs
+		}
+	}
+}
+
+func getServiceByLabel(client clientset.Clientset, service v1.Service) v1.Service{
+	// TODO: query service by service-spec-hash
+	svc, err := client.Core().Services(service.Namespace).List(&unversioned.LabelSelector{MatchLabels: map[string]string{serviceHashLabelName: "baz"}})
+	// One service in one cluster with same hash
+	if err == nil && svc != nil && len(svc.Items) == 0 {
+		return svc.Items[0]
+	} else {
+		return nil
+	}
+}
+
+func (s *ServiceController) ensureClusterServices(service *v1.Service, client clientset.Clientset, clusterName string) ([]string, error) {
+	var err error
+	for i := 0; i < clientRetryCount; i++ {
+		svc := getServiceByLabel(client, service)
+		if svc != nil {
+			// service exists
+			glog.V(4).Infof("service exists in cluster %v, trying to update", clusterName)
+			_, err = client.Core().Services(service.Namespace).UpdateStatus(service)
```

Comment:
Need to check whether underlying service spec is correct, and update as necessary.  We should be able to check against the cached service, to avoid an unnecessary update against the cluster.